arbor/include/arbor/util/hash_def.hpp

```diff
@@ -17,16 +17,46 @@
  */
 
 #include <cstddef>
-#include <typeindex>
+#include <string_view>
 
 // Helpers for forming hash values of compounds objects.
 
 namespace arb {
 
-inline std::size_t hash_value_combine(std::size_t n) {
-    return n;
+// Non-cryptographic hash function for mapping strings to internal
+// identifiers. Concretely, FNV-1a hash function taken from
+//
+//   http://www.isthe.com/chongo/tech/comp/fnv/index.html
+//
+// NOTE: It may be worth it considering different hash functions in
+//       the future that have better characteristic, xxHash or Murmur
+//       look interesting but are more complex and likely require adding
+//       external dependencies.
+//       NOTE: this is the obligatory comment on a better hash function
+//             that will be here until the end of time.
+
+template <typename T>
+inline constexpr std::size_t internal_hash(T&& data) {
```

Comment:
Thanks for moving it here!
I just have one suggestion.
Calling `internal_hash` and `hash_value` would give different results for a single type due to the multiplication with the prime number in `hash_value_combine`. How about merging the `internal_hash` function into `hash_value`? 
So something like this:
```C++
template <typename T>
inline constexpr std::size_t hash_value(const T& t) {
    if constexpr (std::is_convertible_v<T, std::string_view>) {
        constexpr std::size_t prime = 0x100000001b3;
        constexpr std::size_t offset_basis = 0xcbf29ce484222325;

        std::size_t hash = offset_basis;

        for (uint8_t byte: std::string_view{t}) {
            hash = hash ^ byte;
            hash = hash * prime;
        }

        return hash;
    }
    else { return std::hash<std::decay_t<T>>{}(t); }
}

template <typename T, typename... TAIL>
inline constexpr std::size_t hash_value(const T& t, const TAIL&... tail) {
    constexpr std::size_t prime = 93481;
    return prime * internal_hash(t) + hash_value(tail...);
}
```
This way we would simply always use `hash_value` throughout the code.

Reply:
Also, a comment mentioning the different behaviour for `char*` and `const char*` types would be good. For these, the content they point to is hashed. Any other type of pointer will have the hash of the pointer value itself.
Or we do not allow usage with other types of pointers.

Reply:
I think this is a decent solution for the problem of how to treat different pointer types.
The first comment is not yet addressed however. Calling `hash_value` and `internal_hash` for a single type will yield a different hash. As suggested above, we could just overload the `hash_value` function, such that the current `internal_hash` function will just become the `hash_value` function for single types. This way we can also get rid of `hash_value_combine`, which appears to be only used here.

Reply:
Yeah, I missed that change. I think just having `hash_value` would still be overall cleaner then `internal_hash`, `hash_combine` and `hash_value`. But from a functionality point of view I'm ok with it as is.

Reply:
Well, the _surface_ now just is `hash_value`. I should adjust the usage accordingly and stuff `internal_hash` and `hash_combine` into a `detail` namespace.